File: algorithm-interview/ch10/27/gina0605_merge_k_sorted_lists.cpp
/*
23. Merge k Sorted Lists
24ms, 13.8MB
*/

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

typedef pair<int, int> pii;

#define fi first
#define se second

class Solution {
public:
    ListNode* mergeKLists(vector<ListNode*>& lists) {
		priority_queue<pii, vector<pii>, greater<pii>> pq;
		ListNode *head, *tail;
		head = tail = new ListNode();
		for(int i = 0; i < lists.size(); i++) if(lists[i] != nullptr)
			pq.push(pii(lists[i]->val, i));
		while(!pq.empty()) {
			pii t = pq.top();
			pq.pop();
			tail->next = new ListNode(t.fi, nullptr);
			tail = tail->next;
			ListNode* &l = lists[t.se];
			if(l->next != nullptr) {
				l = l->next;
				pq.push(pii(l->val, t.se));
			}
		}
		return head->next;
    }
};

File: algorithm-interview/ch10/27/gina0605_merge_k_sorted_lists.py
'''
23. Merge k Sorted Lists
108ms, 18MB
'''


# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
	def mergeKLists(self, lists: List[ListNode]) -> ListNode:
		pq = []
		head = tail = ListNode()
		for i in range(len(lists)):
			if lists[i]:
				heapq.heappush(pq, (lists[i].val, i))
		while pq:
			(v, i) = heapq.heappop(pq)
			tail.next = ListNode(v)
			tail = tail.next
			lists[i] = lists[i].next
			if lists[i]:
				heapq.heappush(pq, (lists[i].val, i))
		return head.next
